src/Interpreters/sortBlock.cpp

```diff
@@ -35,13 +58,22 @@ ColumnsWithSortDescriptions getColumnsWithSortDescription(const Block & block, c
             ? block.getByName(description[i].column_name).column.get()
             : block.safeGetByPosition(description[i].column_number).column.get();
 
-        res.emplace_back(ColumnWithSortDescription{column, description[i], isColumnConst(*column)});
+        if (isCollationRequired(description[i]))
+        {
+            if (!column->isCollationSupported())
+                throw Exception(
+                    "Collations could be specified only for String, LowCardinality(String), Nullable(String) or for Array or Tuple, "
+                    "containing them.",
+                    ErrorCodes::BAD_COLLATION);
+        }
+        if (const auto * tuple = typeid_cast<const ColumnTuple *>(column))
+            flattenTupleColumnRecursively(res, tuple, &description[i], isColumnConst(*column));
```

Comment:
I am not sure if we need to make a lot of changes.
Currently for each column in ColumnTuple we can create own ColumnSortDescription just copy input ColumnSortDescription, remove Collation if needed (for non String types), and then other code will work as expected.